apps/sms/views/shared/js/activity_handler.js

```diff
@@ -396,37 +399,37 @@ var ActivityHandler = {
             data: { id, threadId }
           };
 
-          var notification = new Notification(title, options);
-          notification.addEventListener('click', goToMessage);
-          releaseWakeLock();
-
-          // Close notification if we are already in thread view and view become
-          // visible.
-          if (document.hidden && threadId === Threads.currentId) {
-            document.addEventListener('visibilitychange',
-              function onVisible() {
-                document.removeEventListener('visibilitychange', onVisible);
-                notification.close();
-            });
-          }
+          return titlePromise.then((title) => {
+            var notification = new Notification(title, options);
+            notification.addEventListener('click', goToMessage);
+            releaseWakeLock();
+
+            // Close notification if we are already in thread view and view
+            // become visible.
+            if (document.hidden && threadId === Threads.currentId) {
+              document.addEventListener('visibilitychange',
+                function onVisible() {
+                  document.removeEventListener('visibilitychange', onVisible);
+                  notification.close();
+              });
+            }
+          });
         }
 
-        function getTitleFromMms(callback) {
+        function getTitleFromMms() {
           // If message is not downloaded notification, we need to apply
           // specific text in notification title;
           // If subject exist, we display subject first;
           // If the message only has text content, display text context;
           // If there is no subject nor text content, display
           // 'mms message' in the field.
+          var defer = Utils.Promise.defer();
+
           if (needManualRetrieve) {
-            setTimeout(function notDownloadedCb() {
-              callback(_('notDownloaded-title'));
-            });
+            formatValue('notDownloaded-title').then(defer.resolve);
           }
           else if (message.subject) {
-            setTimeout(function subjectCb() {
```

Comment:
I think it's because we want to make sure the callback function for `getTitleFromMms` will be executed asynchronously, otherwise it might cause some unexpected problem if callback runs sync in certain case. Now we have promise and we can write the async pattern easily.